Migrate op to new infra: argmax

shutovilyaep opened the pull request:
### What's changed
Migrated the `argmax` device operation from the old vector-based structure to the new TMP device operation pattern. This migration follows the comprehensive guide in `ttnn/cursor/DEVICE_OPERATION_MIGRATION_GUIDE.md`.

**Key changes:**
- **Replaced old device operation** (`argmax_op.cpp/hpp`) with new TMP structure:
  - `argmax_device_operation.hpp` - Main device operation struct following TMP pattern
  - `argmax_device_operation.cpp` - Implementation with static methods
  - `argmax_device_operation_types.hpp` - Type definitions (`operation_attributes_t`, `tensor_args_t`, return types)
- **Updated program factory** (`argmax_program_factory.cpp/hpp`) to work with the new TMP structure
- **Updated public API** (`argmax.cpp/hpp`) to use the registered prim instead of direct device operation calls
- **Maintained backward compatibility** - No changes to the public API signature or behavior

**Technical details:**
- Created `operation_attributes_t` with operation configuration (output_dtype, dim, keepdim, sub_core_grids, use_multicore, output_mem_config)
- Created `tensor_args_t` with tensor parameters (input, optional_output_tensor)
- Implemented `select_program_factory` to choose appropriate program factory variant
- Implemented validation methods (`validate_on_program_cache_hit`, `validate_on_program_cache_miss`)
- Registered prim in `ttnn::prim` namespace
- Updated all call sites to use the registered prim



### Checklist
- [X] [All post commit](https://github.com/tenstorrent/tt-metal/actions/runs/19905825973)
- [X] Code analysis https://github.com/tenstorrent/tt-metal/actions/runs/19907181485

shutovilyaep committed:
refactor(ttnn): Migrate ArgMax device operation to TMP pattern

Migrate the ArgMax reduction operation from the old vector-based
device operation structure to the new TMP (Template Metaprogramming)
pattern that eliminates unnecessary heap allocations.

Changes:
- Add argmax_device_operation_types.hpp with operation_attributes_t,
  tensor_args_t, and return type aliases
- Add argmax_device_operation.hpp/cpp with ArgMaxDeviceOperation struct
  implementing TMP static methods and prim registration
- Refactor argmax_program_factory.hpp/cpp to use TMP factories with
  shared_variables_t and override_runtime_arguments methods
- Update argmax.cpp to use ttnn::prim::argmax instead of operation::run
- Remove legacy argmax_op.hpp/cpp files

ayerofieiev-tt commented on ttnn/cpp/ttnn/operations/reduction/argmax/device/argmax_program_factory.hpp:
Please move each factory into its own hpp/cpp

shutovilyaep replied:
fixed: split argmax_program_factory.hpp/.cpp into separate files:
- argmax_single_core_program_factory.hpp/.cpp
- argmax_multi_core_program_factory.hpp/.cpp

ayerofieiev-tt commented on ttnn/cpp/ttnn/operations/reduction/argmax/device/argmax_device_operation.hpp:
```diff
@@ -0,0 +1,103 @@
+ * @param keepdim Whether to keep the reduced dimension.
+ * @return The output shape.
+ */
+inline ttnn::SmallVector<uint32_t> get_output_shape(
```
Looks like a good function to be in cpp only

shutovilyaep replied:
fixed with *:
- IIUC, "function in cpp only" is not an option due to usage in 2 cpp files (argmax.cpp and argmax_device_operation.cpp).
- Moved implementation from header to argmax_device_operation.cpp, kept declaration in header.

shutovilyaep committed:
fix: Remove const from operation_attributes_t and change Tensor& to Tensor in tensor_args_t

shutovilyaep committed:
fix: Move get_output_shape implementation from header to cpp file

shutovilyaep committed:
fix: Update zero_volume_argmax signature to match get_output_shape

shutovilyaep committed:
fix: Update ArgMaxOperation::invoke signature for const correctness

- Changed dim to const std::optional<int>& (not a simple type)
- Removed const from simple types (keepdim, use_muticore) passed by value
- Kept const& for object parameters

shutovilyaep committed:
fix: Clean up and sort includes in argmax files

- Removed unused includes (core.hpp, run_operation.hpp, etc.)
- Sorted includes according to project conventions (local/project/std)
- Applied clang-format formatting changes

shutovilyaep committed:
fix: split argmax program factories into separate files

shutovilyaep committed:
fix: refactor single-core helper functions with early returns

shutovilyaep commented on ttnn/cpp/ttnn/operations/reduction/argmax/device/argmax_device_operation_types.hpp:
```diff
@@ -0,0 +1,30 @@
+};
+
+struct tensor_args_t {
+    const Tensor& input;
```
fixed: changed const Tensor& to Tensor in tensor_args_t

shutovilyaep committed:
fix: Update copyright from Tenstorrent Inc. to Tenstorrent AI ULC.

Updated all argmax-related files to use 'Tenstorrent AI ULC.' instead of 'Tenstorrent Inc.'

shutovilyaep commented on ttnn/cpp/ttnn/operations/reduction/argmax/argmax.hpp:
```diff
@@ -1,20 +1,18 @@
-// SPDX-FileCopyrightText: © 2023 Tenstorrent Inc.
+// SPDX-FileCopyrightText: © 2025 Tenstorrent Inc.
```
Please note:
- During Copyright text update, I've looked through the code and found two consecutive lines (no empty // string between them), and done same fix in this PR
```
// SPDX-FileCopyrightText: © 2025 Tenstorrent AI ULC.
// SPDX-License-Identifier: Apache-2.0
```
- Later noticed that LLM tends to generate empty // comment between them, there are files with empty comment line, also noticed that some developers included that empty line approach in their Cursor rules/commands
- Kind proposal is to make this 'not required for code understanding, thrown away automatically by brain' copyright-related comments shorter, without empty line between them

shutovilyaep committed:
Fix: using namespace

shutovilyaep committed:
FIx: using OP's own namespace operations::reduction::argmax

shutovilyaep committed:
Fix: reveted early exist, explicit switch

shutovilyaep committed:
fix: error logging improved

shutovilyaep commented on ttnn/cpp/ttnn/operations/reduction/argmax/device/argmax_device_operation.cpp:
```diff
@@ -0,0 +1,184 @@
+        const auto& input_shape = input_tensor_a.padded_shape();
+        auto rank = input_shape.size();
+        // With TILE layout, padded shape has always at least 2 dims (i.e., also for 1D input tensors).
+        TT_FATAL(rank > 1, "Invalid rank for input tensor with TILE layout");
```
fixed

shutovilyaep commented on ttnn/cpp/ttnn/operations/reduction/argmax/device/argmax_single_core_program_factory.cpp:
```diff
@@ -0,0 +1,204 @@
+
+namespace ttnn::operations::reduction::program {
+
+using namespace tt::constants;
```
fixed

shutovilyaep commented on ttnn/cpp/ttnn/operations/reduction/argmax/device/argmax_single_core_program_factory.cpp:
```diff
@@ -0,0 +1,204 @@
+
+    // TILE layout
+    TT_FATAL(
+        output.layout() == Layout::ROW_MAJOR, "For input tensor with TILE layout only ROW_MAJOR output is supported");
```
fixed: using explicit switch

shutovilyaep committed:
fix: auto*

aliaksei-sala commented on ttnn/cpp/ttnn/operations/reduction/argmax/device/argmax_multi_core_program_factory.hpp:
```diff
@@ -0,0 +1,34 @@
+namespace ttnn::operations::reduction::argmax::program {
+
+struct ArgMaxMultiCoreSharedVariables {
+    tt::tt_metal::KernelHandle reader_kernel_id0;
```
Clang will complain: error: constructor does not initialize these fields: reader_kernel_id, ... [cppcoreguidelines-pro-type-member-init,hicpp-member-init,-warnings-as-errors]


You can use the job : https://github.com/tenstorrent/tt-metal/actions/workflows/code-analysis.yaml
to see what else need to be fixed before the merge

aliaksei-sala commented on ttnn/cpp/ttnn/operations/reduction/argmax/device/argmax_device_operation.hpp:
```diff
@@ -0,0 +1,65 @@
+ * @param keepdim Whether to keep the reduced dimension.
+ * @return The output shape.
+ */
+ttnn::SmallVector<uint32_t> get_output_shape(const Tensor& input_tensor, const std::optional<int>& dim, bool keepdim);
```
The header potentially can be public API, let's try to avoid exposing functions in detail namespace. 
We can move such functions into separate file

shutovilyaep committed:
Fix: default initializers, clang-tidy

shutovilyaep committed:
fix: get_output_shape to argmax_utils.hpp

shutovilyaep committed:
fix: API parameter name typo fix use_muticore -> use_multicore

shutovilyaep commented:
/codeowners ping
